backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs

```diff
@@ -40,27 +55,104 @@ public async Task<ICollection<ChallengePreviewDto>> GetChallengesAsync(Challenge
                     challenge.Versions.Any(version => version.LanguageId == filters.LanguageId));
             }
 
-            if (filters.TagsIds != null)
+            if (filters.Progress.HasValue)
             {
-                challenges = challenges.Where(challenge =>
-                    challenge.Tags.Any(tag => filters.TagsIds.Contains(tag.Id)));
+                challenges = FilterChallengesByProgress(challenges, filters.Progress);
             }
 
-            if (filters.Progress.HasValue)
+            if (page is not null && page.PageSize > 0 && page.PageNumber > 0)
+            {
+                challenges = challenges.Skip(page.PageSize * (page.PageNumber - 1))
+                    .Take(page.PageSize);
+            }
+
+            IEnumerable<Challenge> filterChallenges = await challenges.ToListAsync();
+
+            //filter runs on the client because filters.TagsIds[] didn't pass to the server and SQL query get error
+            if (filters.TagsIds != null)
```

Comment:
all filtering should be a part of query (in SQL side)

Reply:
fixed